### PR TITLE
Factor out a utility for replacing types during canonicalization

### PR DESCRIPTION
**Pull request by tlively**
Equirecursive canonicalization generates new minimal type definitions for each
distinct type, but then it must replace the old, non-minimial definitions with
the new ones. That second part where the types are replaced is not unique to
equirecursive canonicalization; the same replacement happens with BasicKind
types and in the future the hybrid isorecursive system will also need to do that
kind of replacement. To improve code reuse and separation of concerns, factor
the type replacement logic out into a separate utility.

This change slightly slows down shape canonicalization, but shape
canonicalization is still much faster than global canonicalization. Nominal
typing performance is not affected.

### CODE REVIEW

File: src/wasm/wasm-type.cpp [hunk 1]
```diff
@@ -2357,6 +2358,114 @@ void TypeBuilder::setNominal(size_t i) {
 
 namespace {
 
+struct CanonicalizationState {
+  // The list of types being built and canonicalized. Will eventually be
+  // returned to the user TypeBuilder user.
```
Comment (kripken) [inline src/wasm/wasm-type.cpp:2363]:
```suggestion
  // returned to the TypeBuilder user.
```

File: src/wasm/wasm-type.cpp [hunk 2]
```diff
@@ -2357,6 +2358,114 @@ void TypeBuilder::setNominal(size_t i) {
 
 namespace {
 
+struct CanonicalizationState {
```
Comment (kripken) [inline src/wasm/wasm-type.cpp:2361]: Maybe add a comment that this is a helper struct for TypeBuilder specifically?

File: src/wasm/wasm-type.cpp [hunk 3]
```diff
@@ -2357,6 +2358,114 @@ void TypeBuilder::setNominal(size_t i) {
...
+};
+
+void CanonicalizationState::update(ReplacementMap& replacements) {
+  if (replacements.empty()) {
+    return;
+  }
+
+  // Update the results vector.
+  for (auto& type : results) {
+    if (auto it = replacements.find(type); it != replacements.end()) {
```
Comment (kripken) [inline src/wasm/wasm-type.cpp:2413]:
```suggestion
    auto it = replacements.find(type);
    if (it != replacements.end()) {
```
perhaps? I don't think we use `;` inside if anywhere in the codebase atm. (I actually didn't know that was possible...)

Comment (tlively) [reply src/wasm/wasm-type.cpp:2413]: This is a new C++17 that we should start using. It avoids `it` being in-scope unnecessarily after the `if {}`.

Comment (kripken) [reply src/wasm/wasm-type.cpp:2413]: I see. A downside though is the line is longer which seems less readable. And in this case we don't need to limit the scope. But I don't feel strongly, if this is common in c++17 then sgtm.

File: src/wasm/wasm-type.cpp (hunk 2 above)
Comment (tlively) [reply src/wasm/wasm-type.cpp:2363]: Done!

File: src/wasm/wasm-type.cpp (hunk 1 above)
Comment (tlively) [reply src/wasm/wasm-type.cpp]: Done!

**Review by kripken — APPROVED** [review]

**Pull request merged**